batch/batch/aioclient.py

```diff
@@ -152,36 +287,31 @@ def __init__(self, client, id, attributes):
         if pvc_size:
             doc['pvc_size'] = pvc_size
 
-        j = await self._client._post('/jobs/create', json=doc)
+        self._job_docs.append(doc)
 
-        return Job(self,
-                   j['job_id'],
-                   attributes=j.get('attributes'),
-                   parent_ids=j.get('parent_ids', []))
+        j = Job.unsubmitted_job(self, self._job_idx, attributes, parent_ids)
+        self._jobs.append(j)
+        return j
 
-    async def close(self):
-        await self._client._patch(f'/batches/{self.id}/close')
+    async def submit(self):
+        if self._submitted:
+            raise ValueError("cannot submit an already submitted batch")
+        self._submitted = True
 
-    async def cancel(self):
-        await self._client._patch(f'/batches/{self.id}/cancel')
+        if len(self._job_docs) != 0:
+            self._doc['jobs'] = self._job_docs
```

Comment:
This doesn't seem to be needed; just check if len(self._job_docs) and `_post(json=self._job_docs)`

Reply:
I need this line because the schema of the spec includes a field 'jobs' that is the array of job specs.

Reply:
I see. I don't actually see where ['jobs'] is read:

@routes.post('/batches/create')
@routes.post('/batches/create')
@authenticated_users_only
async def create_batch(request, userdata):
    parameters = await request.json()

    schema = {
        'attributes': {
            'type': 'dict',
            'keyschema': {'type': 'string'},
            'valueschema': {'type': 'string'}
        },
        'callback': {'type': 'string'},
        'ttl': {'type': 'number'}
    }
    validator = cerberus.Validator(schema)
    if not validator.validate(parameters):
        abort(400, 'invalid request: {}'.format(validator.errors))

    batch = await Batch.create_batch(attributes=parameters.get('attributes'),
                                     callback=parameters.get('callback'),
                                     ttl=parameters.get('ttl'),
                                     userdata=userdata)
    return jsonify(await batch.to_dict(include_jobs=False))
```

Reply:
The code should be this now:
```python3
@routes.post('/batches/create')
@authenticated_users_only
async def create_batch(request, userdata):
    parameters = await request.json()

    validator = cerberus.Validator(schemas.batch_schema)
    if not validator.validate(parameters):
        abort(400, 'invalid request: {}'.format(validator.errors))

    batch = await Batch.create_batch(attributes=parameters.get('attributes'),
                                     callback=parameters.get('callback'),
                                     userdata=userdata)

    job_parameters = parameters.get('jobs', [])
    for job_params in job_parameters:
        await create_job(batch.id, userdata, job_params)
    return jsonify(await batch.to_dict(include_jobs=False))


async def _get_batch(batch_id, user):
    batch = await Batch.from_db(batch_id, user)
    if not batch:
        abort(404)
    return await batch.to_dict(include_jobs=True)
``` 

where schemas.batch_schema is

```python3
batch_schema = {
    'attributes': {
        'type': 'dict',
        'keyschema': {'type': 'string'},
        'valueschema': {'type': 'string'}
    },
    'callback': {'type': 'string'},
    'jobs': {'type': 'list', 'schema': job_schema}
}
```

Reply:
Sweet!